src-control/Control/RefCount.hs

```diff
@@ -525,33 +541,36 @@ assertNoDoubleRelease Ref { reftracker = RefTracker refid _weak outer allocSite
 -- Note however that this is not the only place where 'RefNeverReleased'
 -- exceptions can be thrown. All Ref operations poll for forgotten refs.
 --
-checkForgottenRefs :: IO ()
-checkForgottenRefs = do
+checkForgottenRefs :: forall m. (PrimMonad m, MonadThrow m) => m ()
+checkForgottenRefs =
 #ifndef NO_IGNORE_ASSERTS
     return ()
 #else
-    -- The hope is that by combining `performMajorGC` with `yield` that the
-    -- former starts the finalizer threads for all dropped weak references and
-    -- the latter suspends the current process and puts it at the end of the
-    -- thread queue, such that when the current process resumes the finalizer
-    -- threads for all dropped weak references have finished.
-    -- Unfortunately, this relies on the implementation of the GHC scheduler,
-    -- not on any Haskell specification, and is therefore both non-portable and
-    -- presumably rather brittle. Therefore, for good measure, we do it twice.
-    performMajorGCWithBlockingIfAvailable
-    yield
-    performMajorGCWithBlockingIfAvailable
-    yield
-    assertNoForgottenRefs
-  where
+    throwIfJust =<< do
+    unsafeIOToPrim $ do
+      -- The hope is that by combining `performMajorGC` with `yield` that the
+      -- former starts the finalizer threads for all dropped weak references and
+      -- the latter suspends the current process and puts it at the end of the
+      -- thread queue, such that when the current process resumes the finalizer
+      -- threads for all dropped weak references have finished.
+      -- Unfortunately, this relies on the implementation of the GHC scheduler,
+      -- not on any Haskell specification, and is therefore both non-portable and
+      -- presumably rather brittle. Therefore, for good measure, we do it twice.
+      performMajorGCWithBlockingIfAvailable
+      yield
+      performMajorGCWithBlockingIfAvailable
+      yield
```

Comment:
We could just wrap this bit in `unsafeIOToPrim` and leave `assertNoForgottenRefs` in an outer context that is `(PrimMonad m, MonadThrow m) => m ()`.

Reply:
Done